packages/flutter/lib/src/rendering/mouse_tracking.dart

```diff
@@ -140,19 +188,101 @@ class _MouseState {
   }
 }
 
-/// Maintains the relationship between mouse devices and
-/// [MouseTrackerAnnotation]s, and notifies interested callbacks of the changes
-/// thereof.
+/// Used by [MouseTracker] to provide the details of an update of a mouse
+/// device.
+///
+/// This class contains the information needed to handle the update that might
+/// change the state of a mouse device, or the [MouseTrackerAnnotation]s that
+/// the mouse device is hovering.
+@immutable
+class MouseTrackerUpdateDetails with Diagnosticable {
+  /// When device update is triggered by a new frame.
+  ///
+  /// All parameters are required.
+  const MouseTrackerUpdateDetails.byNewFrame({
+    @required this.lastAnnotations,
+    @required this.nextAnnotations,
+    @required this.previousEvent,
+  }) : assert(previousEvent != null),
+       assert(lastAnnotations != null),
+       assert(nextAnnotations != null),
+       triggeringEvent = null;
+
+  /// When device update is triggered by a pointer event.
+  ///
+  /// The [lastAnnotations], [nextAnnotations], and [triggeringEvent] are
+  /// required.
+  const MouseTrackerUpdateDetails.byPointerEvent({
+    @required this.lastAnnotations,
+    @required this.nextAnnotations,
+    this.previousEvent,
+    @required this.triggeringEvent,
+  }) : assert(triggeringEvent != null),
+       assert(lastAnnotations != null),
+       assert(nextAnnotations != null);
+
+  /// The annotations that the device is hovering before the update.
+  ///
+  /// It is never null.
+  final LinkedHashSet<MouseTrackerAnnotation> lastAnnotations;
+
+  /// The annotations that the device is hovering after the update.
+  ///
+  /// It is never null.
+  final LinkedHashSet<MouseTrackerAnnotation> nextAnnotations;
+
+  /// The last event that the device observed before the update.
+  ///
+  /// If the update is triggered by a frame, it is not null, since the pointer
+  /// must have been added before. If the update is triggered by an event,
```

Comment:
Is this only null for specific events (e.g. pointer added)? Maybe document that?

Reply:
I changed it to 
```dart
  /// The last event that the device observed before the update.
  ///
  /// If the update is triggered by a frame, the [previousEvent] is never null,
  /// since the pointer must have been added before.
  ///
  /// If the update is triggered by a pointer event, the [previousEvent] is not
  /// null except for cases where the event is the first event observed by the
  /// pointer (which is not necessarily a [PointerAddedEvent]).
```